roachtest: disable or fix failing azure tests

DarrylWong opened the pull request:
In the near future we will start running nightly roachtests on Azure. This PR attempts to disable or fix most roachtests that fail on Azure.

The vast majority of failures are due to some sort of cloud incompatibility, i.e. trying to write to a gcs bucket or using google pub/sub changefeeds. In these cases, the test was disabled for Azure. A few tests however, failed due to a lack of setup and were fixed instead.

We also make a change to when the local cluster cache can delete clusters. Instead, we now attempt to clear the cache of deleted clusters at the start of a roachtest run, but do not allow deletion after that. This is due to Azure returning stale information about which VMs we have running. With the original cache behavior, this would see us removing all Azure VMs after `azure.List` returned no VMs.

Release note: none
Fixes: none
Epic: none

DarrylWong committed:
roachtest: add azure configs for tpccbench and tpcc/headroom tests

These tests depend on a per cloud config to determine
the number of warehouses to load in. The numbers were
based off of gce but in the future we should tune the
number after we have more data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121709)
<!-- Reviewable:end -->

DarrylWong commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -908,25 +912,29 @@ func registerTPCC(r registry.Registry) {
+		EstimatedMaxAzure:   900,
 
-		Clouds: registry.AllExceptAWS,
+		Clouds: registry.OnlyGCE,
```
These tests can't run on local, so this should've been `OnlyGCE` in the first place.

Not sure if we want to run all the benchmarks on azure for cost reasons, but doing so would require kicking off a run to figure out better numbers for azure. The current numbers may or may not be high enough from my limited testing.

DarrylWong committed:
roachtest: disable certain cdc sink tests for azure

Several CDC tests rely on streaming to a pubsub or
cloudstorage sink. On azure, we do not have the permissions
currently to do this so we disable them on azure.

DarrylWong committed:
roachtest: disable backup tests on azure

Most backup tests are currently not set up for azure.

DarrylWong committed:
roachtest: disable disk stall tests for azure

Some roachtests simulate disk stalls by using dmsetup
or cgroupsv2. Configs need to be added to Azure for both.
However, the former runs into "Device or resource busy"
and the latter does not properly stall the disk, so
both are disabled for azure until fixed.

DarrylWong committed:
roachtest: fix ORM tests for azure

Some tests are not compatible with azure due
to requiring port service registration.

DarrylWong committed:
roachtest: disable tests that use fixture buckets on azure

These tests require a fixture bucket which is only configured
on gce right now.

DarrylWong commented:
I think this is finally RFAL, `unknown cluster` error seems to be finally fixed. I'll run another nightly over the weekend to make sure nothing unexpected is failing, but in general we should just be disabling failing tests.

renatolabs commented on pkg/roachprod/roachprod.go:
```diff
@@ -237,7 +237,18 @@ func CachedCluster(name string) (*cloud.Cluster, bool) {
+// cache when we know no one is using it, i.e. the start of a roachtest
+// run or a roachprod CLI invocation. This approach is best effort, it
+// is still possible for the caller to concurrently call roachprod sync
+// while a roachtest suite is running.
```
I wonder if we could scope this behaviour _just_ for the Azure nightly run? Maybe via a command line flag that is only passed in that build.

It would help us scope this change and also make this a little more understandable. Currently, some calls get `true` and others get `false`, and it's not super clear to me why one would pass one or the other, especially in unrelated clouds.

DarrylWong replied:
We chatted offline and agreed this was a good idea. I went with the `roachtest` setting a public var in `roachprod approach. It seemed slightly easier to follow than an env var.

renatolabs reviewed (commented):
Main comment for your consideration is the handling of `overwriteMissingClusters`. Nice work!

DarrylWong committed:
roachtest: add clear-cluster-cache flag

This flag indicates if we should remove missing clusters
from the local cache when syncing resources with providers.

Because Azure can sometimes return stale VM information
when many PUT calls are made in succession, we cannot
trust that missing clusters in ListCloud are actually
deleted.

This flag allows azure runs that spin up multiple
concurrent clusters, i.e. azure nightlies, to
disable clearing the cache.

DarrylWong commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -351,11 +351,15 @@ var tpccSupportedWarehouses = []struct {
 	{hardware: "gce-n4cpu16", v: version.MustParse(`v2.1.0-0`), warehouses: 1300},
 	{hardware: "gce-n4cpu16", v: version.MustParse(`v19.1.0-0`), warehouses: 1250},
 	{hardware: "aws-n4cpu16", v: version.MustParse(`v19.1.0-0`), warehouses: 2100},
+	// TODO(darrylwong): this number is copied from gce-n4cpu16. Adjust once we have a
```
Hmm, I actually don't remember what I did here, it doesn't match any of the numbers lol. I just know that the numbers "pass". 

Considering the values haven't been touched since 2019, I'll replace it with a more generic "look at all the numbers again" TODO.

DarrylWong commented on pkg/cmd/roachtest/tests/backup.go:
```diff
@@ -314,14 +314,17 @@ func initBulkJobPerfArtifacts(testName string, timeout time.Duration) (func(), *
+		// to 400 on GCE, which is not enough for this test. We could request a
+		// larger volume size and set spec.LocalSSDDisable if we wanted to run
+		// this on Azure.
+		CompatibleClouds:  registry.OnlyGCE,
```
Yeah it wasn't working for me. I suppose another way to get this work on Azure would be to reduce the number of rows, but then it wouldn't be `backup/2TB` anymore 😅

DarrylWong commented on pkg/cmd/roachtest/tests/acceptance.go:
```diff
@@ -132,7 +133,7 @@ func registerAcceptance(r registry.Registry) {
 				EncryptionSupport: tc.encryptionSupport,
 				Timeout:           10 * time.Minute,
-				CompatibleClouds:  registry.AllExceptAWS,
+				CompatibleClouds:  registry.AllExceptAWS.Remove(tc.incompatibleClouds...),
```
Agreed, I don't see why any of them besides the multitenant ones wouldn't work on AWS anyway.

DarrylWong commented on pkg/cmd/roachtest/tests/multitenant_tpch.go:
```diff
@@ -155,7 +156,7 @@ func registerMultiTenantTPCH(r registry.Registry) {
 				Benchmark:        true,
 				Cluster:          r.MakeClusterSpec(1 /* nodeCount */),
-				CompatibleClouds: registry.AllExceptAWS,
+				CompatibleClouds: registry.Clouds(spec.Local, spec.GCE), // Requires service registration.
```
Good idea, done.

renatolabs reviewed (approved):
LGTM!

DarrylWong commented:
TFTR!

bors r=renatolabs

srosenberg commented on pkg/roachprod/roachprod.go:
```diff
@@ -232,6 +232,13 @@ func CachedCluster(name string) (*cloud.Cluster, bool) {
 
+// ClearClusterCache indicates if we should ever clear the local cluster
+// cache of clusters. This flag is set to false during Azure nightly runs,
+// as the large amount of concurrent resources created will cause Azure.List
```
Yikes 😱

srosenberg reviewed (commented):
Nice work!
